homeassistant/components/pushover/notify.py

```diff
@@ -46,67 +51,87 @@ def __init__(self, hass, user_key, api_token):
         self._hass = hass
         self._user_key = user_key
         self._api_token = api_token
-        self.pushover = Client(self._user_key, api_token=self._api_token)
+        self.pushover = PushoverAPI(self._api_token)
 
     def send_message(self, message="", **kwargs):
         """Send a message to a user."""
-        # Make a copy and use empty dict if necessary
-        data = dict(kwargs.get(ATTR_DATA) or {})
-
-        data["title"] = kwargs.get(ATTR_TITLE, ATTR_TITLE_DEFAULT)
 
-        # Check for attachment.
-        if ATTR_ATTACHMENT in data:
+        # Extract params from data dict
+        title = kwargs.get(ATTR_TITLE, ATTR_TITLE_DEFAULT)
+        data = dict(kwargs.get(ATTR_DATA) or {})
+        url = data.get(ATTR_URL, None)
+        url_title = data.get(ATTR_URL_TITLE, None)
+        priority = data.get(ATTR_PRIORITY, None)
+        retry = data.get(ATTR_PRIORITY, None)
+        expire = data.get(ATTR_EXPIRE, None)
+        callback_url = data.get(ATTR_CALLBACK_URL, None)
+        timestamp = data.get(ATTR_TIMESTAMP, None)
+        sound = data.get(ATTR_SOUND, None)
+        html = 1 if data.get(ATTR_HTML, False) else 0
+
+        image = data.get(ATTR_ATTACHMENT, None)
+        # Check for attachment
+        if image is not None:
             # If attachment is a URL, use requests to open it as a stream.
-            if data[ATTR_ATTACHMENT].startswith("http"):
+            if image.startswith("http"):
                 try:
                     response = requests.get(
                         data[ATTR_ATTACHMENT], stream=True, timeout=5
```

Comment:
Remove this. We should not allow Home Assistant to download URLs and then send it out in a notification. Sending out urls is fine. Sending out from a whitelisted path is fine too.

Reply:
Hmm, I'm fine with removing it, I'm not using it anyways.

But this was also possible before.

And I think it is not possible to call from outside or without authentication, so I guess it's not harmful? What about sending a picture from a camera with the notification? I guess it would be useful to download urls then?

Reply:
It was possible before but it will be removed from all notification integrations. It's dangerous because the Home Assistant machine can now download anything, including pages that are localhost only and send it to another computer.

Reply:
Agreed. Now it only checks agains whitelisted path and refuse other attachments.